execute/tokendata/observer_background.go

```diff
@@ -0,0 +1,431 @@
+func (o *backgroundObserver) Observe(
+	_ context.Context,
+	observations exectypes.MessageObservations,
+) (exectypes.TokenDataObservations, error) {
+	o.lggr.Debug("Observe called",
+		"observations", observations,
+		"cachedTokenData", o.cachedTokenData.size(),
+		"queuedMsgs", o.msgQueue.size(),
+	)
+
+	tokenDataResults := make(exectypes.TokenDataObservations)
+
+	// initialize the response with non-ready data since we have to respond with the same structure and
+	// number of elements as the input
+	for chainSel, seqNumToMsg := range observations {
+		tokenDataResults[chainSel] = make(map[cciptypes.SeqNum]exectypes.MessageTokenData)
+		for seqNum, msg := range seqNumToMsg {
+			tokenDataResults[chainSel][seqNum] = exectypes.MessageTokenData{
+				TokenData: make([]exectypes.TokenData, len(msg.TokenAmounts)),
+			}
+
+			for i := range msg.TokenAmounts {
+				tokenDataResults[chainSel][seqNum].TokenData[i] = exectypes.TokenData{
+					Ready:     false,
+					Supported: o.IsTokenSupported(chainSel, msg.TokenAmounts[i]),
+				}
+			}
+		}
+	}
+
+	// override with the cached data that are ready
+	for chainSel, seqNumToMsg := range observations {
+		for seqNum, msg := range seqNumToMsg {
+			tokenData, exists := o.cachedTokenData.get(msg.Header.MessageID)
+			if exists && !tokenData.SupportedAreReady() {
+				return nil, fmt.Errorf("internal error, cache contains not ready token data")
+			}
+
+			lggr := logger.With(o.lggr, "msgID", msg.Header.MessageID.String(),
+				"sourceChain", chainSel.String(), "seqNum", seqNum.String())
+
+			if exists {
+				// token data exist so include them in the results
+				lggr.Infow("token data found in cache")
+				tokenDataResults[chainSel][seqNum] = tokenData
+			} else {
+				// token data not in cache for this message, enqueue the message
+				if ok := o.msgQueue.enqueue(msg, 0); ok {
+					lggr.Infow("message added to the queue")
+				} else {
+					lggr.Infow("message already exists in the queue")
+				}
+			}
+		}
+	}
+
+	return tokenDataResults, nil
+}
+
+// IsTokenSupported simply forwards the call to the underlying observer.
+func (o *backgroundObserver) IsTokenSupported(
+	sourceChain cciptypes.ChainSelector, msgToken cciptypes.RampTokenAmount) bool {
+	return o.observer.IsTokenSupported(sourceChain, msgToken)
+}
+
+// Close stops the background goroutines and cleans up resources.
+func (o *backgroundObserver) Close() {
+	close(o.done)
+	o.wg.Wait()
+}
+
+// startWorkers starts the worker goroutines that process messages from the queue.
+func (o *backgroundObserver) startWorkers() {
+	o.lggr.Info("waiting for existing (if any) workers to stop")
+	o.wg.Wait()
+	o.lggr.Info("all workers stopped, new workers are starting...")
+
+	for i := 0; i < o.numWorkers; i++ {
+		o.wg.Add(1)
+		workerID := i
+		go o.worker(workerID)
+	}
+}
+
+// worker is a goroutine that processes messages from the queue.
+// It will stop using the o.done channel.
+func (o *backgroundObserver) worker(id int) {
```

Comment:
I think we miss QoS/ isolation for processing between different integrations. Let's imagine 99% of traffic is from USDC and USDC is down for 1 hour. However, we keep retrying USDC messages leading into starvation of "Integration B" messages because of keeping them in the same queue. 

I think it could be solved by either:
* partitioning / having a queue per integration 
* having `backgroundWorker` per tokenDataObserver instead of global one